gson/src/main/java/com/google/gson/TypeAdapter.java

```diff
@@ -287,4 +292,105 @@ public final T fromJsonTree(JsonElement jsonTree) {
+   * It is recommended that callers of this method wrap thrown exceptions and add
+   * context information, such as {@link JsonReader#getPath()}, to them.
+   *
+   * <p>The default implementation of this method calls {@link #read(JsonReader)}
+   * with a JsonReader containing only the property name as {@link JsonToken#STRING}.
+   *
+   * <p>Adapter implementations which explicitly want to support deserialization from
+   * a property name should override this method, and can for example delegate to a
+   * private parsing method which they also use for {@link #read(JsonReader)}.
+   * Adapter implementations which do <i>not</i> want to support deserialization
+   * from a property name may override this method to throw an
+   * {@link UnsupportedOperationException}.
+   *
+   * <p>For serialization using Gson's built-in {@code Map} adapter, an alternative
+   * to overriding this method can be to use {@link GsonBuilder#enableComplexMapKeySerialization()}.
+   *
+   * @param name the JSON property name; not {@code null}
+   * @return Java object (possibly {@code null}) parsed from the property name
+   * @throws UnsupportedOperationException If deserialization from a property
+   *        name is not supported by this adapter
+   * @throws JsonParseException If parsing the property name fails
+   */
+  // TODO: Throwing UnsupportedOperationException is not ideal because it 'bubbles up'
+  // too far the call stack, but Gson has no better fitting exception which also
+  // fits for createPropertyName
+  public T readFromPropertyName(String name) throws JsonParseException {
+    JsonTreeReader jsonReader = new PropertyNameJsonReader(name);
+    boolean wasLenient = jsonReader.isLenient();
+    T value;
+    try {
+      value = read(jsonReader);
+    } catch (IOException e) {
+      throw new JsonParseException("Failed reading from name", e);
+    }
+    // Catch RuntimeException to cover all exceptions which could occur during parsing
+    catch (RuntimeException e) {
+      throw new JsonParseException("Failed reading from name", e);
+    }
+
+    if (jsonReader.peek() != JsonToken.END_DOCUMENT) {
+      throw new IllegalStateException("Adapter did not consume name");
+    }
+    if (wasLenient != jsonReader.isLenient()) {
+      // Throw exception to allow detecting cases where type adapter might expect
+      // that change affects subsequent deserialization (which is not the case here)
+      throw new IllegalStateException("Lenient setting was changed but not reverted");
+    }
+    return value;
+  }
+
+  static class PropertyNameJsonReader extends JsonTreeReader {
+    public PropertyNameJsonReader(String propertyName) {
+      super(new JsonPrimitive(propertyName));
+    }
+
+    @Override public String getPath() {
+      // Does not have access to actual path, so at least make it obvious that
+      // path is unknown
+      return "#fake-property-name-path";
```

Comment:
This path and especially the "fake" might be a bit irritating for users; maybe just use `<property-name>` or similar as path? That hopefully should make it obvious enough that this is not the actual name.